convert2rhel/grub.py

```diff
@@ -0,0 +1,597 @@
+        raise BootloaderError("Cannot get required information about the EFI partition.")
+    logger.debug("Block device: %s" % str(blk_dev))
+    logger.debug("ESP device number: %s" % str(dev_number))
+
+    efi_path = None
+    for filename in DEFAULT_INSTALLED_EFIBIN_FILENAMES:
+        tmp_efi_path = os.path.join(RHEL_EFIDIR_CANONICAL_PATH, filename)
+        if os.path.exists(tmp_efi_path):
+            efi_path = canonical_path_to_efi_format(tmp_efi_path)
+            logger.debug("The new EFI binary: %s" % tmp_efi_path)
+            logger.debug("The new EFI binary [efi format]: %s" % efi_path)
+            break
+    if not efi_path:
+        BootloaderError("Cannot detect any RHEL EFI binary file.")
+
+    cmd_fmt = "/usr/sbin/efibootmgr -c -d %s -p %s -l '%s' -L '%s'"
+    cmd_params = (blk_dev, dev_number["minor"], efi_path, label)
+
+    stdout, ecode = utils.run_subprocess(cmd_fmt % cmd_params, print_output=False)
+    if ecode:
+        logger.debug("efibootmgr output:\n-----\n%s\n-----" % stdout)
+        raise BootloaderError("Cannot create the new EFI bootloader entry for RHEL.")
+
+    # check that our entry really exists, if yes, it will be default for sure
+    logger.info("Check the new EFI bootloader.")
+    new_efibootinfo = EFIBootInfo()
+    new_boot_entry = None
+    for i in new_efibootinfo.entries.values():
+        if i.label == label and efi_path in i.efi_bin_source:
+            new_boot_entry = i
+    if not new_boot_entry:
+        raise BootloaderError("Cannot get the boot number of the new EFI bootloader entry.")
+
+
+def _remove_current_boot_entry(efibootinfo_orig):
+    """Remove the current (original) boot entry if ...
+
+    if:
+        - the referred EFI binary file doesn't exist anymore and originally
+          was located in the default directory for the original OS
+        - the referred EFI binary file is same as the current default one
+
+    The conditions could be more complicated if algorithms in this module
+    are changed. Additional checks are implemented that could prevent the
+    current boot from the removal.
+
+    The function is expected to be called after the new RHEL entry is created.
+    but expects to get EFIBootInfo before the new bootloader entry is created.
+    """
+    efibootinfo_new = EFIBootInfo()
+    # the following checks are just preventive. the warnings should not appear,
+    # unless someone changes the code in future providing a bug
+    orig_boot = efibootinfo_new.entries.get(efibootinfo_orig.current_boot, None)
+    if not orig_boot:
+        logger.warning(
+            "The original default EFI bootloader entry %s has been removed already."
+            % orig_boot.boot_number
+        )
+        return
+    if orig_boot != efibootinfo_orig.entries[orig_boot.boot_number]:
+        logger.warning(
+            "The EFI bootloader entry %s has been modified. Skipping the removal."
+            % orig_boot.boot_number
+        )
+        return
```

Comment:
@pirat89 @bocekm I've setup a CentOS 8 with UEFI and either the VM is setup incorrectly or the creation of a boot entry is incorrect and would want to hear your thoughts.

Before creation we have this. You'll notice there isn't an entry for `0001`, so when we remove CentOS (`0004`) and add in a new entry for RHEL, it gets assigned `0001`. When we then do the check to make sure they're different we make an assumption it will be in the same spot
```
[root@localhost ~]# /usr/sbin/efibootmgr
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0002,0000,0003
Boot0000* UiApp
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
Boot0004* CentOS Linux
```

Should we instead keep track of the new bootloader number and compare that instead of assuming it will be the same number as the CentOS one? Or is this handling a case for an unexpected configuration? Or am I completely misunderstanding what is going on? :D

This is the logs with changes to bootloader
```
[...]
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
Remove the original EFI boot entry '0004'.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -Bb 0004'

[09/09/2021 12:14:45] TASK - [Convert: Patch yum configuration file] ****************************                 
[09/09/2021 12:14:45] DEBUG - Calling command 'rpm -Vf /etc/dnf/dnf.conf'
Skipping patching, yum configuration file not modified

[09/09/2021 12:14:45] TASK - [Final: rpm files modified by the conversion] **********************                 
Skipping the execution of 'rpm -Va'.
Skipping comparison of the 'rpm -Va' output from before and after the conversion.

[09/09/2021 12:14:45] TASK - [Final: Configure the bootloader] **********************************                 
The new expected EFI binary found: /boot/efi/EFI/redhat/shimx64.efi
Copy the GRUB2 configuration files to the new EFI directory.
The /boot/efi/EFI/redhat/grubenv file already exists. Copying skipped.
The /boot/efi/EFI/redhat/grub.cfg file already exists. Copying skipped.
The original /boot/efi/EFI/centos directory is kept. Remove the directory manually after you check it's not needed anymore.
Load the bootloader configuration.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0001,0002,0000,0003
Boot0000* UiApp
Boot0001* Red Hat Enterprise Linux 8
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -v'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0001,0002,0000,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
Boot0002* UEFI Misc Device      PciRoot(0x0)/Pci(0x2,0x3)/Pci(0x0,0x0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
Replace the current EFI bootloader entry with the RHEL one.
Create the 'Red Hat Enterprise Linux 8' EFI bootloader entry.
[09/09/2021 12:14:45] DEBUG - Calling command 'lsblk -spnlo MAJ:MIN /dev/vda1'
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - Calling command 'lsblk -spnlo name /dev/vda1'
[09/09/2021 12:14:45] DEBUG - Block device: /dev/vda
[09/09/2021 12:14:45] DEBUG - ESP device number: {'major': 252, 'minor': 1}
[09/09/2021 12:14:45] DEBUG - The new EFI binary: /boot/efi/EFI/redhat/shimx64.efi
[09/09/2021 12:14:45] DEBUG - The new EFI binary [efi format]: \EFI\redhat\shimx64.efi
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -c -d /dev/vda -p 1 -l '\EFI\redhat\shimx64.efi' -L 'Red Hat Enterprise Linux 8''
Check the new EFI bootloader.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp
Boot0001* Red Hat Enterprise Linux 8
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
Boot0004* Red Hat Enterprise Linux 8
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -v'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
Boot0002* UEFI Misc Device      PciRoot(0x0)/Pci(0x2,0x3)/Pci(0x0,0x0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
Boot0004* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp
Boot0001* Red Hat Enterprise Linux 8
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
Boot0004* Red Hat Enterprise Linux 8
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -v'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
Boot0002* UEFI Misc Device      PciRoot(0x0)/Pci(0x2,0x3)/Pci(0x0,0x0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
Boot0004* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
[09/09/2021 12:14:45] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 110, in main
    grub.post_ponr_set_efi_configuration()
  File "/usr/lib/python3.6/site-packages/convert2rhel/grub.py", line 590, in post_ponr_set_efi_configuration
    _replace_efi_boot_entry(efibootinfo)
  File "/usr/lib/python3.6/site-packages/convert2rhel/grub.py", line 525, in _replace_efi_boot_entry
    _remove_current_boot_entry(efibootinfo)
  File "/usr/lib/python3.6/site-packages/convert2rhel/grub.py", line 471, in _remove_current_boot_entry
    if orig_boot != efibootinfo_orig.entries[orig_boot.boot_number]:
KeyError: '0004'
```

Reply:
I'd not rely on specific boot entry position (#4 in this case). Check/track what's position of the entry added by us.

Reply:
@SpyTec  @bocekm the comparison in this one case has a different meaning. (see the warning msg in the body of the condition). This one specific check is supposed to get info whether the original entry is still same or not. In case you change position, it's not original entry. Entries are not numbered strictily in order. You can have set a bootloader with entries ('0000', '0003', '0004', '0006', ..). Let's call it identifier instead of the position, this makes the additional thinking about that easier. The condition is not supposed to check a newly added entry, but the original one. That's why the numbering is strict. 

The piece of the code for the context
```python
    if orig_boot != efibootinfo_orig.entries[orig_boot.boot_number]:            
        logger.warning(                                                         
            "The EFI bootloader entry %s has been modified. Skipping the removal."
            % orig_boot.boot_number                                             
        )
```
Where:
- `orig_boot` is supposed to be the original boot entry after the conversion and bootloader update (new bootloader entry has been added already). read: bootloader entry with the same boot number after the recent changes (if exists)
- you want to compare it with the bootloader entry before the conversion.

Unless the code has been updated, I do not understand why the keyerror happens, as from the shared output, clearly the entry `0004` existed, so it cannot miss in the `efibootinfo_orig` dict.

Reply:
Thanks for the info Petr. I let it remain but just made sure that the entry exists when the check is being made

I also made a check to see if a RHEL boot entry already exists, so we shouldn't get duplicated entries

Boot order will likely be correct, so did not create any methods for verifying and correcting it